<a href="https://colab.research.google.com/github/janaeam/food-sales-predictions/blob/main/Data_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Getting to Know My Data



In [23]:
import pandas as pd

In [24]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [25]:
filename = '/content/drive/MyDrive/Coding Dojo/Project 1/sales_predictions.csv'
sales = pd.read_csv(filename)

In [26]:
sales.head()

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0           FDA15         9.30  ...  Supermarket Type1          3735.1380
1           DRC01         5.92  ...  Supermarket Type2           443.4228
2           FDN15        17.50  ...  Supermarket Type1          2097.2700
3           FDX07        19.20  ...      Grocery Store           732.3800
4           NCD19         8.93  ...  Supermarket Type1           994.7052

[5 rows x 12 columns]

In [27]:
# Are the dtypes consistent with the presumed column category

sales.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

In [28]:
# Which columns have missing values?

sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Item_Weight and Outlet_Size columns are both missing data

I will explore further below

# Checking each column for unique value counts

In [29]:
sales['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [30]:
sales.loc[ : , 'Item_Fat_Content'].replace( {'LF' : 'Low Fat',
                                            'low fat' : 'Low Fat',
                                             'reg' : 'Regular'}, inplace=True)

In [31]:
sales['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

In [32]:
sales['Item_Identifier'].value_counts()

FDG33    10
FDW13    10
FDP25     9
FDV38     9
FDX31     9
         ..
DRF48     1
FDQ60     1
FDK57     1
FDE52     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64

In [33]:
sales['Item_MRP'].value_counts()

172.0422    7
188.1872    6
170.5422    6
109.5228    6
196.5084    6
           ..
212.8218    1
190.3872    1
162.6868    1
189.1214    1
51.3008     1
Name: Item_MRP, Length: 5938, dtype: int64

In [34]:
sales['Item_Outlet_Sales'].value_counts()

958.7520     17
1342.2528    16
1845.5976    15
703.0848     15
1278.3360    14
             ..
3167.8764     1
2226.4352     1
1684.4740     1
1574.6170     1
6692.6216     1
Name: Item_Outlet_Sales, Length: 3493, dtype: int64

In [35]:
sales['Item_Type'].value_counts()

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

In [36]:
sales['Item_Weight'].value_counts()

12.150    86
17.600    82
13.650    77
11.800    76
15.100    68
          ..
7.560      2
9.420      1
5.400      1
6.520      1
7.685      1
Name: Item_Weight, Length: 415, dtype: int64

In [51]:
sales['Outlet_Establishment_Year'].value_counts()

1985    1463
1987     932
1999     930
1997     930
2004     930
2002     929
2009     928
2007     926
1998     555
Name: Outlet_Establishment_Year, dtype: int64

In [38]:
sales['Outlet_Identifier'].value_counts()

OUT027    935
OUT013    932
OUT049    930
OUT046    930
OUT035    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: Outlet_Identifier, dtype: int64

In [39]:
sales['Outlet_Location_Type'].value_counts()

Tier 3    3350
Tier 2    2785
Tier 1    2388
Name: Outlet_Location_Type, dtype: int64

In [40]:
sales['Outlet_Size'].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [41]:
sales['Outlet_Type'].value_counts()

Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: Outlet_Type, dtype: int64

## Columns to Drop

I will drop only the "Item_Visibility" column for now. Initially I thought this column would prove to be useful in predicting sales baased on the visibility of each individual product, however, since the lowest visibility products, including those with zero visibility, account for the majority of the dataset (77%). There is no need to qualify this coulumn's data values beecause it is irrelevent. 
Also, the products with the highest visibiilty, both above .3 and .2, account for only 1.4%, further proving this column's irrelevance.

#### It is my belief that we can assume the product's visibility value has very little to do with it's sale value.

In [42]:
vis_filter = sales['Item_Visibility'].count()
vis_filter1 = (sales['Item_Visibility'] >= .0) & (sales['Item_Visibility'] < .1)
vis_filter2 = (sales['Item_Visibility'] >= .1) & (sales['Item_Visibility'] < .2)
vis_filter3 = (sales['Item_Visibility'] >= .2) & (sales['Item_Visibility'] < .3)
vis_filter4 = sales['Item_Visibility'] >= .3

vis_count1 = sales.loc[vis_filter1, 'Item_Visibility'].count()
vis_count2 = sales.loc[vis_filter2, 'Item_Visibility'].count()
vis_count3 = sales.loc[vis_filter3, 'Item_Visibility'].count()
vis_count4 = sales.loc[vis_filter4, 'Item_Visibility'].count()

print(vis_count1)
print(round(vis_count1/vis_filter, 2))
print(vis_count2)
print(round(vis_count2/vis_filter, 2))
print(vis_count3)
print(round(vis_count3/vis_filter, 3))
print(vis_count4)
print(round(vis_count4/vis_filter, 4))

6593
0.77
1796
0.21
122
0.014
12
0.0014


In [43]:
sales.drop(columns = 'Item_Visibility', inplace = True)

## Exploring the relevance of missing data in the Item_Weight and Outlet_Size columns

In [53]:
outlet_missing = sales['Outlet_Size'].isna()
sales.loc[outlet_missing, :].head()

Item_Identifier  Item_Weight  ...        Outlet_Type Item_Outlet_Sales
3            FDX07       19.200  ...      Grocery Store          732.3800
8            FDH17       16.200  ...  Supermarket Type1         1076.5986
9            FDU28       19.200  ...  Supermarket Type1         4710.5350
25           NCD06       13.000  ...  Supermarket Type1          838.9080
28           FDE51        5.925  ...      Grocery Store          178.4344

[5 rows x 11 columns]

In [45]:
sales[outlet_missing].shape

(2410, 11)

#### Since all of the missing data for the Outlet_Size is specific to these Outlet_Identifier numbers we will learn nothing about what the missing values should be.

In [103]:
# Checking value counts with the outlet_missing filter on the Outlet_Identifier column to see how 
# the missing Outlet_Size values compare to the Outlet_Identifier value counts already checked above

sales.loc[outlet_missing, 'Outlet_Identifier'].value_counts()

OUT045    929
OUT017    926
OUT010    555
Name: Outlet_Identifier, dtype: int64

In [105]:
# Checking how the Outlet_Type and Outlet_Identifier columns compare together with the missing
# values from the Outlet_Size column

sales.loc[outlet_missing, ['Outlet_Identifier', 'Outlet_Type']].value_counts()

Outlet_Identifier  Outlet_Type      
OUT045             Supermarket Type1    929
OUT017             Supermarket Type1    926
OUT010             Grocery Store        555
dtype: int64

#### In the above cell where the value counts for the Outlet_Type column "Supermarket Type1" accounted for 5577 cells I will check to see if all the other Outlet_Type rows with "Supermarket Type1" contain a consistent Outlet_Size as this will allow me to see what the Outlet_Size values should be.

In [107]:
# Checking value counts with the outlet_missing filter on the Outlet_Type column to see how 
# the missing Outlet_Size values compare to the Outlet_Type value counts already checked above

sales.loc[outlet_missing, 'Outlet_Type'].value_counts()

Supermarket Type1    1855
Grocery Store         555
Name: Outlet_Type, dtype: int64

In [120]:
supType1_filter = sales['Outlet_Type'] == 'Supermarket Type1'
sales.loc[supType1_filter, 'Outlet_Size'].value_counts()

Small     1860
High       932
Medium     930
Name: Outlet_Size, dtype: int64

In [145]:
Groc_filter = sales['Outlet_Type'] == 'Grocery Store'
size = sales['Outlet_Size'] == 'Small'
sales[Groc_filter & ~outlet_missing].count()

Item_Identifier              528
Item_Weight                    0
Item_Fat_Content             528
Item_Type                    528
Item_MRP                     528
Outlet_Identifier            528
Outlet_Establishment_Year    528
Outlet_Size                  528
Outlet_Location_Type         528
Outlet_Type                  528
Item_Outlet_Sales            528
dtype: int64

In [151]:
sales.loc[Groc_filter, 'Outlet_Size'].value_counts()

Small    528
Name: Outlet_Size, dtype: int64

In [152]:
sales.loc[size, 'Outlet_Type'].value_counts()

Supermarket Type1    1860
Grocery Store         528
Name: Outlet_Type, dtype: int64

In [153]:
sales.loc[supType1_filter, 'Outlet_Size'].value_counts()

Small     1860
High       932
Medium     930
Name: Outlet_Size, dtype: int64

In [141]:
sales[Groc_filter & outlet_missing].count()

Item_Identifier              555
Item_Weight                  555
Item_Fat_Content             555
Item_Type                    555
Item_MRP                     555
Outlet_Identifier            555
Outlet_Establishment_Year    555
Outlet_Size                    0
Outlet_Location_Type         555
Outlet_Type                  555
Item_Outlet_Sales            555
dtype: int64

In [144]:
OUT045_filter = sales['Outlet_Identifier'] == 'OUT045'
OUT017_filter = sales['Outlet_Identifier'] == 'OUT017'
OUT010_filter = sales['Outlet_Identifier'] == 'OUT010'

sales[Groc_filter & OUT045_filter | OUT017_filter | OUT010_filter].count()

Item_Identifier              1481
Item_Weight                  1481
Item_Fat_Content             1481
Item_Type                    1481
Item_MRP                     1481
Outlet_Identifier            1481
Outlet_Establishment_Year    1481
Outlet_Size                     0
Outlet_Location_Type         1481
Outlet_Type                  1481
Item_Outlet_Sales            1481
dtype: int64

In [46]:
weight_missing = sales['Item_Weight'].isna()
sales.loc[weight_missing, :].head()

Item_Identifier  Item_Weight  ...        Outlet_Type Item_Outlet_Sales
7            FDP10          NaN  ...  Supermarket Type3         4022.7636
18           DRI11          NaN  ...  Supermarket Type3         2303.6680
21           FDW12          NaN  ...  Supermarket Type3         4064.0432
23           FDC37          NaN  ...      Grocery Store          214.3876
29           FDC14          NaN  ...      Grocery Store          125.8362

[5 rows x 11 columns]

In [47]:
sales[weight_missing].shape

(1463, 11)

## Removing Unwanted Data Observations

In [48]:
# Checking for duplicates

sales.duplicated().any()

False